### PR TITLE
Fix #5 - Add XKCD

### LINKED ISSUE #5: xkcd

### PR DESCRIPTION
**Pull request by Turakar**

### CODE REVIEW

**Comment by rleh** [comment]
Please use meaningful pull request and branch names and commit messages :wink:

**Comment by rleh** [comment]
:+1:

File: src/main/resources/de/nerdclubtfg/signalbot/components/defaultConfig.json [hunk 1]
```diff
@@ -2,7 +2,8 @@
   "plugins": {
     "echo": true,
     "fefe" : true,
-    "plugins" : true
+    "plugins" : true,
```
Comment (rleh) [inline src/main/resources/de/nerdclubtfg/signalbot/components/defaultConfig.json:5]: Adding a trailing comma would produce nicer diffs.
Con: [This does not conform to JSON.org](http://stackoverflow.com/questions/201782/can-you-use-a-trailing-comma-in-a-json-object) / [RRF4627](http://www.ietf.org/rfc/rfc4627.txt),
but: [FasterXML/Jackson supports trailing commas](https://github.com/FasterXML/jackson-core/issues/118).

Comment (Turakar) [reply src/main/resources/de/nerdclubtfg/signalbot/components/defaultConfig.json:5]: I had problems with GSON (google framework for json) generating null entries in the HashMap. Is this also true for FasterXML?

Comment (rleh) [reply src/main/resources/de/nerdclubtfg/signalbot/components/defaultConfig.json:5]: @Turakar ?

Comment (Turakar) [reply src/main/resources/de/nerdclubtfg/signalbot/components/defaultConfig.json:5]: Jackson 2.8.5 (latest) does not seem to support this

File: src/main/java/de/nerdclubtfg/signalbot/plugins/Xkcd.java [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+	private InputStream open(String url) throws IOException {
+		HttpsURLConnection connection = (HttpsURLConnection) new URL(URL + url).openConnection();
+		InputStream in = connection.getInputStream();
+		return in;
+	}
+	
+	private File toFile(String url) throws IOException {
+		HttpsURLConnection connection = (HttpsURLConnection) new URL(url.replace("http://", "https://")).openConnection();
+		InputStream in = connection.getInputStream();
+		File temp = File.createTempFile("xkcd", ".png");
```
Comment (rleh) [inline src/main/java/de/nerdclubtfg/signalbot/plugins/Xkcd.java:91]: Da Java kein Multi-threading unterstützt kann bei `xkcd.png` keine Race Condition entstehen, oder?
Ich fände es schöner wenn die xkcd-id mit im Dateinamen steht. Oder ein zufällig generierter Dateiname?

File: src/main/resources/de/nerdclubtfg/signalbot/components/defaultConfig.json (hunk 1 above)
Comment (rleh) [reply src/main/resources/de/nerdclubtfg/signalbot/components/defaultConfig.json:5]: ok

**Review by rleh — CHANGES_REQUESTED** [review]

File: src/main/java/de/nerdclubtfg/signalbot/plugins/Xkcd.java (hunk 2 above)
Comment (Turakar) [reply src/main/java/de/nerdclubtfg/signalbot/plugins/Xkcd.java:91]: Java creates "xkcd[somethingrandom].png from that in the folder /tmp or the windows equivalent. This ensures that the user/os knows that this file can be deleted after termination of the bot.

File: src/main/java/de/nerdclubtfg/signalbot/plugins/Xkcd.java [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+	public Xkcd() {
+		super("!xkcd");
+	}
+
+	@Override
+	public void onMessage(User sender, Group group, SignalServiceDataMessage message) throws IOException {
+		try {
+			String body = stripPrefix(message.getBody().get());
+			if(!body.isEmpty()) {
+				sendXkcd(sender, group, body);
```
Comment (rleh) [inline src/main/java/de/nerdclubtfg/signalbot/plugins/Xkcd.java:41]: Better:

``` java
if(body.matches("\\d+")) {
    sendXkcd(sender, group, body);
} else {
    Signal.getInstance().sendMessage(sender, group, "Error: invalid xkcd id.");
}

```

**Review by rleh — CHANGES_REQUESTED** [review]

File: src/main/java/de/nerdclubtfg/signalbot/plugins/Xkcd.java (hunk 3 above)
Comment (Turakar) [reply src/main/java/de/nerdclubtfg/signalbot/plugins/Xkcd.java:41]: fixed

File: src/main/java/de/nerdclubtfg/signalbot/plugins/Xkcd.java (hunk 2 above)
Comment (rleh) [reply src/main/java/de/nerdclubtfg/signalbot/plugins/Xkcd.java:91]: Ok :+1:

**Review by rleh — APPROVED** [review]

**Pull request merged**